doc/pekwm_dialog.1
.nh
.TH pekwm\_dialog 1 "February 2021" pekwm\_dialog "User Manual"

.SH NAME
.PP
pekwm\_dialog \- a xmessage inspired dialog application


.SH SYNOPSIS
.PP
pekwm\_dialog [\-\-display] message


.SH DESCRIPTION
.PP
pekwm\_dialog presents a simple dialog with a title, image, message
text and a button for each options specified. If no options are
specified the dialog will have an Ok button.

.PP
The caller can identify which options was selected by checking the
exit code of the application. The first option will exit with status
code 0, next with 1 and so on.

.PP
Given this behavior it is recommended to have ok as the first option
and then cancel as the second and then further options. This caters
for standard exit code 0 on ok, and cancel would give the same exit
code as on errors.

.PP
The illustration below depicts the layout of dialog elements:

.PP
.RS

.nf
\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
| TITLE                                    |
\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
|                                          |
| Image if any is displayed here           |
|                                          |
\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
| Message text goes here                   |
|                                          |
\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
|           [Option1] [Option2]            |
\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

.fi
.RE


.SH OPTIONS
.PP
\fB\-\-help\fP Show help information.

.PP
\fB\-\-display\fP \fIDISPLAY\fP Connect to DISPLAY instead of DISPLAY set in environment.

.PP
\fB\-\-geometry\fP \fIGEOMETRY\fP Geometry of window, default is 200x50 expanding for image size.

.PP
\fB\-\-image\fP \fIIMAGE\fP Path to image to display.

.PP
\fB\-\-option\fP \fISTRING\fP Option title, multiple options are supported.

.PP
\fB\-\-title\fP \fISTRING\fP Title string.


NEWS
NEWS pekwm-0.1.7
================

Bugs fixed

  * #XX Several memory leaks identified with valgrind plugged.
  * #XX Fixed broken iconfied state reported by
     <giuliani{.}blackfenix{@}gmail{.}com>
  * #XX CfgParser parsed data twice.
  * #XX StatusWindow placement fixed.
  * #XX Fix for _NET_WM_NAME ops by Florian Hoppe <flp{@}floppi{.}de>
  * #22 Font type is now possible to define as the first argument in
     themes.
  * #30 Panel raise issues, reported and fixed by Jyri Jokinen
     <shared{@}adresh{.}com>
  * #21 Dynamic menu entries become unselectable after first
     visit. Reported by Lassi Pölönen and fixed by Andreas
     <ioerror{@}lavabit{.}com>
  * #14 SendKey action requested by nidan.
  * #29 drag'n'drop requires the recieving window to be moved before
     working. Reported by Jyri Jokinen <shared{@}adresh{.}com>, fixed
     by Andreas <ioerror{@}lavabit{.}com>
  * #32 Shaping issues after shading. Reported by Paulo Igor Barra
     Nascimento <pibarnas{@}yahoo{.}com{.}br> and fixed by Andreas
     <ioerror{@}lavabit{.}com>
  * #XX Corrects the interpretation of the motif windowmanager hints
     by Andreas <ioerror{@}lavabit{.}com>
  * #XX Remove perl scripts reducing dependencies on packages. 

User visible changes

  * UTF-8 support added for reading and rendering window titles.
  * _NET_WM_ICON is now read and rendered in menus.
  * RandR support for multi-screen configurations.
  * New default theme.
  * CmdDialog size is now calculated from the current head and not the
    whole display.

Configuration changes

  * _NET_WM_WINDOW_TYPE handling configurable via autoproperties.
  * Width/Height 0 in autoproperties now means screen width/height.
  * WorkspacesPerRow and WorkspaceNames added with matching
    GotoWorkspace/SendToWorkspace parameters Up, PrevV, Down, NextV.
  * ShowWorkspaceIndicator
  * WorkspaceIndicatorScale
  * PCRE support remove.
  * EdgeIndent option added to Screen section of main configuration
    file. If true reserves space where the screen edge is.
  * EdgeSize now supports per edge configuration. 

Theme changes

  * Shaping suport for non-corner windows.
  * XMB type is used by default in favour of X11 type fonts.
  * Font alpha is now handles as % from 0-100 instead of previous Xft
    unsigned int value.
  * HeightAdapt option added to theme.
  * Font type is now possible to define as the first argument in themes.


src/pekwm_dialog.cc
//
// pekwm_dialog.cc for pekwm
// Copyright (C) 2021 Claes Nästén <pekdon@gmail.com>
//
// This program is licensed under the GNU GPL.
// See the LICENSE file for more information.
//

#include "pekwm.hh"

#include "CfgParser.hh"
#include "Debug.hh"
#include "FontHandler.hh"
#include "ImageHandler.hh"
#include "TextureHandler.hh"
#include "Theme.hh"
#include "PWinObj.hh"
#include "Util.hh"
#include "x11.hh"

#include <algorithm>
#include <iostream>
#include <iomanip>
#include <vector>
#include <sstream>

extern "C" {
#include <X11/Xutil.h>
#include <getopt.h>
#include <time.h>
#include <unistd.h>
}

static const uint WIDTH_DEFAULT = 250;
static const uint HEIGHT_DEFAULT = 50;
#define THEME_DEFAULT DATADIR "/pekwm/themes/default/theme"

static int _stop = -1;
static FontHandler* _font_handler = nullptr;
static ImageHandler* _image_handler = nullptr;
static TextureHandler* _texture_handler = nullptr;

namespace pekwm
{
    FontHandler* fontHandler()
    {
        return _font_handler;
    }

    ImageHandler* imageHandler()
    {
        return _image_handler;
    }

    TextureHandler* textureHandler()
    {
        return _texture_handler;
    }
}

/**
 * Dialog
 *
 * --------------------------------------------
 * | TITLE                                    |
 * --------------------------------------------
 * | Image if any is displayed here           |
 * |                                          |
 * |                                          |
 * |                                          |
 * --------------------------------------------
 * | Message text goes here                   |
 * |                                          |
 * --------------------------------------------
 * |           [Option1] [Option2]            |
 * --------------------------------------------
 *
 */
class PekwmDialog : public PWinObj {
public:
    class Widget{
    public:
        virtual ~Widget(void)
        {
            if (_window != None) {
                X11::destroyWindow(_window);
            }
        }

        int getX(void) const { return _gm.x; }
        int getY(void) const { return _gm.y; }

        virtual bool setState(Window window, ButtonState state) {
            return false;
        }
        virtual bool click(Window window) { return false; }
        virtual void render(void) = 0;

        virtual void place(int x, int y, uint width)
        {
            _gm.x = x;
            _gm.y = y;
            _gm.width = width;
        }

        /**
         * Get requested width, 0 means adapt to given width.
         */
        virtual uint widthReq(void) const { return 0; }

        /**
         * Get requested height, given the provided width.
         */
        virtual uint heightReq(uint width) const = 0;

    protected:
        Widget(Theme::DialogData* data, PWinObj &parent)
            : _data(data),
              _window(None),
              _parent(parent)
        {
        }

        void setWindow(Window window) { _window = window; }

    protected:
        Theme::DialogData *_data;
        Window _window;
        PWinObj &_parent;
        Geometry _gm;
    };

    class Button : public Widget {
    public:
        Button(Theme::DialogData* data, PWinObj& parent,
               int retcode, const std::wstring& text)
            : Widget(data, parent),
              _retcode(retcode),
              _text(text),
              _font(data->getButtonFont()),
              _state(BUTTON_STATE_FOCUSED)
        {
            XSetWindowAttributes attr;
            attr.background_pixel = X11::getWhitePixel();
            attr.override_redirect = True;
            attr.event_mask =
                ButtonPressMask|ButtonReleaseMask|
                EnterWindowMask|LeaveWindowMask;
            _gm.width = widthReq();
            _gm.height = heightReq(_gm.width);

            setWindow(X11::createWindow(_parent.getWindow(),
                                        0, 0, _gm.width, _gm.height, 0,
                                        CopyFromParent, InputOutput,
                                        CopyFromParent,
                                        CWEventMask|CWOverrideRedirect|
                                        CWBackPixel, &attr));
            X11::mapWindow(_window);
        }

        virtual ~Button(void)
        {
        }

        virtual bool setState(Window window, ButtonState state) override {
            if (window != _window) {
                return false;
            }
            _state = state;
            render();
            return true;
        }

        virtual bool click(Window window) override {
            if (window != _window) {
                return false;
            }
            if (_state == BUTTON_STATE_HOVER
                || _state == BUTTON_STATE_PRESSED) {
                _stop = _retcode;
            }
            return true;
        }

        virtual void place(int x, int y, uint width) override {
            Widget::place(x, y, _gm.width);
            X11::moveWindow(_window, _gm.x, _gm.y);
        }

        virtual uint widthReq(void) const override {
            return _font->getWidth(_text) + _data->padVert();
        }

        virtual uint heightReq(uint width) const override {
            return _font->getHeight() + _data->padHorz();
        }

        virtual void render(void) override {
            X11::clearWindow(_window);
            _data->getButton(_state)->render(_window, 0, 0,
                                            _gm.width, _gm.height);
            _font->setColor(_data->getButtonColor());
            _font->draw(_window,
                        _data->getPad(PAD_LEFT), _data->getPad(PAD_UP),
                        _text);
        }

    private:
        int _retcode;
        std::wstring _text;
        PFont *_font;
        ButtonState _state;
    };

    class ButtonsRow : public Widget
    {
    public:
        ButtonsRow(Theme::DialogData* data, PWinObj& parent,
                   std::vector<std::wstring> options)
            : Widget(data, parent)
        {
            int i = 0;
            for (auto it : options) {
                _buttons.push_back(new Button(_data, parent, i++, it));
            }
        }

        virtual ~ButtonsRow(void)
        {
            for (auto it : _buttons) {
                delete it;
            }
        }

        virtual bool setState(Window window, ButtonState state) override {
            for (auto it : _buttons) {
                if (it->setState(window, state)) {
                    return true;
                }
            }
            return false;
        }

        virtual bool click(Window window) override {
            for (auto it : _buttons) {
                if (it->click(window)) {
                    return true;
                }
            }
            return false;

        }

        virtual void place(int x, int y, uint width) override
        {
            Widget::place(x, y, width);

            // place buttons centered on available width
            uint buttons_width = 0;
            for (auto it : _buttons) {
                buttons_width += it->widthReq();
            }
            buttons_width += _buttons.size() * _data->padHorz();

            x = (width - buttons_width) / 2;
            y += _data->getPad(PAD_UP);
            for (auto it : _buttons) {
                it->place(x, y, width);
                x += it->widthReq() + _data->padHorz();
            }
        }

        virtual uint heightReq(uint width) const override {
            uint height = 0;
            for (auto it : _buttons) {
                uint height_req = it->heightReq(width);
                if (height_req > height) {
                    height = height_req;
                }
            }
            return height + _data->padVert();
        }

        virtual void render(void) override {
            for (auto it : _buttons) {
                it->render();
            }
        }

    private:
        std::vector<Button*> _buttons;
    };

    class Image : public Widget {
    public:
        Image(Theme::DialogData* data, PWinObj& parent, PImage* image)
            : Widget(data, parent),
              _image(image)
        {
        }
        virtual ~Image(void) { }

        virtual uint widthReq(void) const override
        {
            return _image->getWidth();
        }

        virtual uint heightReq(uint width) const override
        {
            // TODO: if width < image width, this should return the
            // scaled image height considering the aspect
            return _image->getHeight();
        }

        virtual void render(void) override
        {
            // render image centered on available width
            uint x = (_gm.width - _image->getWidth()) / 2;
            _image->draw(_parent.getWindow(),
                         x, _gm.y,
                         _image->getWidth(), _image->getHeight());
        }

    private:
        PImage *_image;
    };

    class Text : public Widget {
    public:
        Text(Theme::DialogData* data, PWinObj& parent,
             const std::wstring& text, bool is_title)
            : Widget(data, parent),
              _font(is_title ? data->getTitleFont() : data->getTextFont()),
              _text(text),
              _is_title(is_title)
        {
        }
        virtual ~Text(void) { }

        virtual uint heightReq(uint width) const override
        {
            // TODO: split up in lines
            return _font->getHeight() + _data->padVert();
        }

        virtual void render(void) override
        {
            _font->setColor(_is_title
                            ? _data->getTitleColor() : _data->getTextColor());
            _font->draw(_parent.getWindow(),
                        _gm.x + _data->getPad(PAD_LEFT),
                        _gm.y + _data->getPad(PAD_UP), _text);
        }

    private:
        PFont *_font;
        std::wstring _text;
        bool _is_title;
    };

    PekwmDialog(Theme::DialogData* data,
                const Geometry &gm,
                bool raise, const std::wstring& title, PImage* image,
                const std::wstring& message,
                std::vector<std::wstring> options)
        : PWinObj(true),
          _data(data),
          _raise(raise)
    {
        _gm = gm;

        initWindow(title);
        // TODO: setup size minimum based on image
        initWidgets(title, image, message, options);
        placeWidgets();
    }

    ~PekwmDialog(void)
    {
        for (auto it : _widgets) {
            delete it;
        }
        X11::destroyWindow(_window);
    }

    virtual void resize(uint width, uint height) override
    {
        PWinObj::resize(width, height);
        for (auto it : _widgets) {
            it->place(it->getX(), it->getY(), width);
        }
        render();
    }

    void show(void)
    {
        X11::mapWindow(_window);
        if (_raise) {
            X11::raiseWindow(_window);
        }
        render();
    }

    void render(void)
    {
        X11::clearWindow(_window);
        _data->getBackground()->render(_window, 0, 0, _gm.width, _gm.height);
        for (auto it : _widgets) {
            it->render();
        }
    }

    void setState(Window window, ButtonState state)
    {
        for (auto it : _widgets) {
            if (it->setState(window, state)) {
                break;
            }
        }
    }

    void click(Window window)
    {
        for (auto it : _widgets) {
            if (it->click(window)) {
                break;
            }
        }
    }

protected:
    void initWindow(const std::wstring& title)
    {
        _window =
            X11::createSimpleWindow(X11::getRoot(),
                                    _gm.x, _gm.y, _gm.width, _gm.height, 0,
                                    X11::getBlackPixel(), X11::getWhitePixel());
        X11::selectInput(_window,
                         ExposureMask|StructureNotifyMask);

        auto c_title = new uchar[title.size() + 1];
        memcpy(c_title, title.c_str(), title.size() + 1);

        XSizeHints normal_hints = {0};
        XWMHints wm_hints = {0};
        wm_hints.flags = StateHint|InputHint;
        wm_hints.initial_state = NormalState;
        wm_hints.input = True;

        char wm_name[] = "dialog";
        char wm_class[] = "pekwm_dialog";
        XClassHint class_hint = {wm_name, wm_class};

        auto title_utf8 =
            Util::to_utf8_str(title.size() ? title : L"pekwm_dialog");
        Xutf8SetWMProperties(X11::getDpy(), _window,
                             title_utf8.c_str(), title_utf8.c_str(), 0, 0,
                             &normal_hints, &wm_hints, &class_hint);

        delete [] c_title;

        X11::setAtom(_window, WINDOW_TYPE, WINDOW_TYPE_NORMAL);
    }

    void initWidgets(const std::wstring& title, PImage* image,
                     const std::wstring& message,
                     std::vector<std::wstring> options)
    {
        if (title.size()) {
            _widgets.push_back(new Text(_data, *this, title, true));
        }
        if (image) {
            _widgets.push_back(new Image(_data, *this, image));
        }
        if (message.size()) {
            _widgets.push_back(new Text(_data, *this, message, false));
        }
        _widgets.push_back(new ButtonsRow(_data, *this, options));
    }

    void placeWidgets(void)
    {
        // height is dependent on the available width, get requested
        // width first.
        uint width = _gm.width;
        for (auto it : _widgets) {
            uint width_req = it->widthReq();
            if (width_req && width_req > width) {
                width = width_req;
            }
        }
        if (width == 0) {
            width = WIDTH_DEFAULT;
        }

        uint y = 0;
        for (auto it : _widgets) {
            it->place(0, y, width);
            y += it->heightReq(width);
        }

        PWinObj::resize(std::max(width, _gm.width), std::max(y, _gm.height));
    }

private:
    Theme::DialogData* _data;
    bool _raise;
    std::vector<Widget*> _widgets;
};

static void init(Display* dpy)
{
    _font_handler = new FontHandler();
    _image_handler = new ImageHandler();
    _texture_handler = new TextureHandler();
}

static void cleanup()
{
    delete _texture_handler;
    delete _image_handler;
    delete _font_handler;
}

static void usage(const char* name, int ret)
{
    std::cout << "usage: " << name
              << " [-dhit] [-o option|-o option...] message" << std::endl
              << "  -d --display dpy    Display" << std::endl
              << "  -h --help           Display this information" << std::endl
              << "  -i --image          Image under title" << std::endl
              << "  -o --option         Option (many allowed)" << std::endl
              << "  -t --title          Dialog title" << std::endl;
    exit(ret);
}

static int mainLoop(PekwmDialog& dialog)
{
    XEvent ev;
    while (_stop == -1 && X11::getNextEvent(ev)) {
        switch (ev.type) {
        case ButtonPress:
            dialog.setState(ev.xbutton.window, BUTTON_STATE_PRESSED);
            break;
        case ButtonRelease:
            dialog.click(ev.xbutton.window);
            break;
        case ConfigureNotify:
            dialog.resize(ev.xconfigure.width, ev.xconfigure.height);
            break;
        case EnterNotify:
            dialog.setState(ev.xbutton.window, BUTTON_STATE_HOVER);
            break;
        case Expose:
            dialog.render();
            break;
        case LeaveNotify:
            dialog.setState(ev.xbutton.window, BUTTON_STATE_FOCUSED);
            break;
        case MapNotify:
            break;
        case ReparentNotify:
            break;
        case UnmapNotify:
            break;
        default:
            DBG("UNKNOWN EVENT " << ev.type);
            break;
        }
    }
    return _stop;
}

static void getThemeDir(const std::string& config_file,
                        std::string& dir, std::string& variant)
{
    CfgParser cfg;
    cfg.parse(config_file, CfgParserSource::SOURCE_FILE, true);
    auto files = cfg.getEntryRoot()->findSection("FILES");
    if (files != nullptr) {
        std::vector<CfgParserKey*> keys;
        keys.push_back(new CfgParserKeyPath("THEME", dir, THEME_DEFAULT));
        keys.push_back(new CfgParserKeyString("THEMEVARIANT", variant));
        files->parseKeyValues(keys.begin(), keys.end());
        std::for_each(keys.begin(), keys.end(), Util::Free<CfgParserKey*>());
    } else {
        dir = THEME_DEFAULT;
        variant = "";
    }
}

int main(int argc, char* argv[])
{
    const char* display = NULL;
    Geometry gm = { 0, 0, WIDTH_DEFAULT, HEIGHT_DEFAULT };
    int gm_mask = WIDTH_VALUE | HEIGHT_VALUE;
    bool raise;
    std::wstring title;
    std::string config_file = Util::getEnv("PEKWM_CONFIG_FILE");
    std::string image_name;
    std::vector<std::wstring> options;

    static struct option opts[] = {
        {"config", required_argument, NULL, 'c'},
        {"display", required_argument, NULL, 'd'},
        {"geometry", required_argument, NULL, 'g'},
        {"help", no_argument, NULL, 'h'},
        {"image", required_argument, NULL, 'i'},
        {"option", required_argument, NULL, 'o'},
        {"raise", no_argument, NULL, 'r'},
        {"title", required_argument, NULL, 't'},
        {NULL, 0, NULL, 0}
    };

    try {
        std::locale::global(std::locale(""));
    } catch (const std::runtime_error &e) {
        setlocale(LC_ALL, "");
    }

    Util::iconv_init();

    int ch;
    while ((ch = getopt_long(argc, argv, "c:d:g:hi:o:rt:", opts, NULL)) != -1) {
        switch (ch) {
        case 'c':
            config_file = optarg;
            break;
        case 'd':
            display = optarg;
            break;
        case 'g':
            gm_mask |= X11::parseGeometry(optarg, gm);
            break;
        case 'h':
            usage(argv[0], 0);
            break;
        case 'i':
            image_name = optarg;
            Util::expandFileName(image_name);
            break;
        case 'o':
            options.push_back(Util::to_wide_str(optarg));
            break;
        case 'r':
            raise = true;
            break;
        case 't':
            title = Util::to_wide_str(optarg);
            break;
        default:
            usage(argv[0], 1);
            break;
        }
    }

    if (config_file.empty()) {
        config_file = Util::getConfig("config");
    }
    Util::expandFileName(config_file);

    if (options.empty()) {
        options.push_back(L"Ok");
    }

    std::wstring message;
    for (int i = optind; i < argc; i++) {
        if (i > optind) {
            message += L' ';
        }
        message += Util::to_wide_str(argv[i]);
    }

    auto dpy = XOpenDisplay(display);
    if (! dpy) {
        std::cerr << "Can not open display!" << std::endl
                  << "Your DISPLAY variable currently is set to: "
                  << Util::getEnv("DISPLAY") << std::endl;
        return 1;
    }

    X11::init(dpy, true);
    init(dpy);

    _image_handler->path_push_back("./");
    PImage *image = nullptr;
    if (image_name.size()) {
        image = _image_handler->getImage(image_name);
    }

    std::string theme_dir, theme_variant;
    getThemeDir(config_file, theme_dir, theme_variant);

    int ret;
    {
        // run in separate scope to get Theme destructed before cleanup
        Theme theme(_font_handler, _image_handler, _texture_handler,
                    theme_dir, theme_variant);
        PekwmDialog dialog(theme.getDialogData(),
                           gm, raise, title, image, message, options);
        dialog.show();

        ret = mainLoop(dialog);
    }
    if (image) {
        _image_handler->returnImage(image);
    }

    cleanup();
    X11::destruct();
    Util::iconv_deinit();

    return ret;
}
